src/olympia/signing/views.py

```diff
@@ -78,32 +81,38 @@ class VersionView(APIView):
     permission_classes = [IsAuthenticated]
 
     @handle_read_only_mode
-    @with_addon(allow_missing=True)
-    def put(self, request, addon, version_string):
-        if 'upload' in request.FILES:
-            filedata = request.FILES['upload']
-        else:
-            return Response(
-                {'error': _('Missing "upload" key in multipart file data.')},
-                status=status.HTTP_400_BAD_REQUEST)
+    def post(self, request):
+        version_string = request.data.get('version', None)
 
         try:
-            # Parse the file to get and validate package data with the addon.
-            pkg = parse_addon(filedata, addon)
-            if not acl.submission_allowed(request.user, pkg):
-                raise forms.ValidationError(
-                    _(u'You cannot submit this type of add-on'))
-        except forms.ValidationError as e:
-            return Response({'error': e.message},
-                            status=status.HTTP_400_BAD_REQUEST)
-        if pkg['version'] != version_string:
+            filedata, pkg = self.handle_upload(request, None, version_string)
+        except forms.ValidationError as exc:
             return Response(
-                {'error': _('Version does not match the manifest file.')},
+                {'error': exc.message},
+                status=exc.code or status.HTTP_400_BAD_REQUEST)
+
+        if not pkg.get('guid', None) and not pkg.get('is_webextension', False):
+            msg = _('Only WebExtensions are allowed to omit the GUID')
+            return Response(
+                {'error': _(msg)},
                 status=status.HTTP_400_BAD_REQUEST)
-        elif (addon is not None and
-                addon.versions.filter(version=version_string).exists()):
-            return Response({'error': _('Version already exists.')},
-                            status=status.HTTP_409_CONFLICT)
+
+        addon = Addon.create_addon_from_upload_data(
```

Comment:
this line looks like it's shared by both `self.put()` and `self.post()`. You could move it to `self.handle_upload()` and then return `was_created` so that `self.put()` can set the status code.